Storage API - make public API more consistent

joeyfreund opened the pull request:
This PR build up on the work that was done in #48 - Change `#cleanup` to take an `experiment`, instead of a `scope`.

This is more consistent with the rest of API (public methods deal with experiments, protected methods deal with scopes and keys).

Also, updated some comments that were probably overlooked when copy-pasted.

joeyfreund commented:
@gmcgibbon , before updating the tests, I wanted to get your feedback.      
Specifically, what I'm interested in is:

 1. Do you agree with this change?
 2. Who is already using the public `cleanup` interface?        
     (since it was very recently added, I feel like it might be safe to introduce this backwards in compatible change in the next minor release)

joeyfreund commented on lib/verdict/storage/base_storage.rb:
```diff
@@ -45,11 +45,12 @@ def store_start_timestamp(experiment, timestamp)
 
-      def cleanup(_scope)
-        raise NotImplementedError
+      def cleanup(experiment)
```
> Passing a `scope` can log a deprecation warning.

Sounds like the safe way to deprecate this.

joeyfreund commented on lib/verdict/storage/base_storage.rb:
```diff
@@ -58,21 +59,29 @@ def get(scope, key)
+      # - The scope is provided as string.
+      # - Should return true if all items were successfully removed from storage.
+      # - Should raise Verdict::StorageError if anything goes wrong.
+      def remove_all_keys(scope)
```
> should call this method `clear`

👍 

 > can rename the redis storage cleanup method `scrub`

I'm a bit confused - Are we talking about the public `cleanup` method (that was added in release `0.10`)?

gmcgibbon commented on lib/verdict/storage/base_storage.rb:
```diff
@@ -45,11 +45,18 @@ def store_start_timestamp(experiment, timestamp)
+        if experiment_or_scope.is_a?(Verdict::Experiment)
+          clear(experiment_or_scope.handle.to_s)
+        else
+          puts("TODO: Log a deprecation warning!")
```
We don't have Active Support in this repo so let's use Verdict's default_logger.

```ruby
          Verdict.default_logger.warn(
            "Passing a scope string to #{self.class}.cleanup is deprecated, " \
            "pass a Verdict::Experiment instance instead."
          )
```

Also IMO the `if` should test for the parameter we know is deprecated and pass anything else to the other code path. Was it a string?

joeyfreund replied:
👍 

> Was it a string?

According to tests, it is a symbol. But I'm assuming it could also be a string.

gmcgibbon commented on test/storage/redis_storage_test.rb:
```diff
@@ -71,7 +71,7 @@ def test_started_at
   end
 
-  def test_cleanup
+  def test_cleanup_with_scope_argument
```
We have mocha, so we should be able to assert the deprecation is logged.

joeyfreund replied:
👍 Done

pawandubey commented on lib/verdict/storage/base_storage.rb:
```diff
@@ -45,11 +45,21 @@ def store_start_timestamp(experiment, timestamp)
+      def cleanup(experiment_or_scope)
+        if experiment_or_scope.is_a?(Symbol) || experiment_or_scope.is_a?(String)
+          Verdict.default_logger.warn(
+            "Passing a scope string/symbol to #{self.class}.cleanup is deprecated, " \
```
Nit: 
```suggestion
            "Passing a scope string/symbol to #{self.class}#cleanup is deprecated, " \
```

joeyfreund replied:
Thanks, I will apply this change when rebasing

joeyfreund commented:
@pawandubey / @gmcgibbon , I just rebased and squashed the commits, with a changelog note.

Can one of you please give it a quick look (just to confirm that I'm conforming to whatever conventions we have in this repo) before I merge this PR?

gmcgibbon commented on CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## Next version
```
Nitpick:
```suggestion
## Unreleased
```
